### PR TITLE
Parse expr bounds checking

### PR DESCRIPTION
**Pull request by chfast**
Requires #158

### CODE REVIEW

File: lib/fizzy/parser.hpp [hunk 1]
```diff
@@ -14,7 +14,13 @@ template <typename T>
 using parser_result = std::tuple<T, const uint8_t*>;
 
 Module parse(bytes_view input);
-parser_result<Code> parse_expr(const uint8_t* input);
+
+parser_result<Code> parse_expr(const uint8_t* input, const uint8_t* end);
+
+inline auto parse_expr(const bytes& input)
```
Comment (chfast) [inline lib/fizzy/parser.hpp]: This can be moved to tests.

File: lib/fizzy/parser_expr.cpp [hunk 2]
```diff
@@ -394,8 +407,7 @@ parser_result<Code> parse_expr(const uint8_t* pos)
         }
         code.instructions.emplace_back(instr);
     }
-    if (!label_positions.empty())
-        throw parser_error{"code is not balanced (missing end statements)"};
+    assert(label_positions.empty());
```
Comment (gumb0) [inline lib/fizzy/parser_expr.cpp]: Why is this changed?

Comment (chfast) [reply lib/fizzy/parser_expr.cpp]: This is not reachable. The parsing ends when the final `end` is reached. Previously it was also not reachable, it is leftover from some earlier jumps implementation. I left assert as a documentation / target for fuzzing.

File: lib/fizzy/parser.hpp (hunk 1 above)
Comment (chfast) [reply lib/fizzy/parser.hpp]: Moved.

File: lib/fizzy/parser_expr.cpp [hunk 3]
```diff
@@ -209,6 +209,9 @@ parser_result<Code> parse_expr(const uint8_t* pos, const uint8_t* end)
 
         case Instr::block:
         {
+            if (pos == end)
```
Comment (axic) [inline lib/fizzy/parser_expr.cpp]: Why not move these to line 43?

Comment (axic) [reply lib/fizzy/parser_expr.cpp]: Only having it in cases where there's an immediate?

Actually, why are we not putting arity onto the immediate list?

File: lib/fizzy/parser_expr.cpp [hunk 4]
```diff
@@ -209,6 +209,9 @@ parser_result<Code> parse_expr(const uint8_t* pos, const uint8_t* end)
...
+                throw parser_error{"Unexpected EOF"};
+
             const uint8_t type{*pos};
```
Comment (axic) [inline lib/fizzy/parser_expr.cpp]: Perhaps it would be due having a helper for the block/valtype because this code is repeated three times: block, loop, if

File: lib/fizzy/parser_expr.cpp [hunk 5]
```diff
@@ -349,15 +349,15 @@ parser_result<Code> parse_expr(const uint8_t* pos, const uint8_t* end)
...
+            std::tie(imm, pos) = leb128s_decode<int32_t>(pos, end);
             push(code.immediates, static_cast<uint32_t>(imm));
             break;
         }
 
         case Instr::i64_const:
         {
             int64_t imm;
-            std::tie(imm, pos) = leb128s_decode<int64_t>(pos, pos + 10);  // FIXME: Bounds checking.
+            std::tie(imm, pos) = leb128s_decode<int64_t>(pos, end);
```
Comment (axic) [inline lib/fizzy/parser_expr.cpp]: Th commit log for this is wrong, as it is not only u32 immediates.

File: test/unittests/parser_expr_test.cpp [hunk 6]
```diff
@@ -204,3 +204,14 @@ TEST(parser, control_instr_out_of_bounds)
     EXPECT_THROW_MESSAGE(parse_expr("03"_bytes), parser_error, "Unexpected EOF");
     EXPECT_THROW_MESSAGE(parse_expr("04"_bytes), parser_error, "Unexpected EOF");
 }
+
+TEST(parser, immediate_u32_out_of_bounds)
```
Comment (axic) [inline test/unittests/parser_expr_test.cpp]: This is not only about u32 immediates.

File: lib/fizzy/parser_expr.cpp (hunk 4 above)
Comment (axic) [reply lib/fizzy/parser_expr.cpp]: Done in #174.
